### ISSUE TITLE
myloader hangs on "waiting for handler commit" / consumes lots of ram

### ISSUE THREAD

**Issue by billynoah** [mydumper/mydumper#1529]
**Describe the bug**
Attempting to restore with myloader causes numerous threads to take very long to complete and consumes so much RAM that eventually the oom killer kills the mysql service.

**To Reproduce**
Command executed:
mydumper --compress=ZSTD --no-locks --threads=8 --chunk-filesize=512 --database=my_db --outputdir=dump_dir
myloader -o -t 4 -B my_db -d dump_dir

What mydumper and myloader version has been used?
myloader v0.16.3-3, built against MySQL 8.0.36-28 with SSL support
mydumper v0.16.3-3, built against MySQL 8.0.36-28 with SSL support

**Log**
If applicable, add the --verbose 3 / -v 3 and --logfile / -L <filename> to the execution of the command and upload the file.

**Environment (please complete the following information):**
 - OS version: Ubuntu 20.04.6

**Additional context**
I realize this may be something unrelated directly to myloader but it has only recently been an issue. I've reviewed this Q&A (https://stackoverflow.com/questions/72029457/mysql-waiting-for-handler-commit) and set `replica_parallel_workers=1` to no avail.  I have a development server that occasionally will be able to complete a restore without crashing but I am still seeing threads marked "waiting for handler commit" that take upwards of 10 minutes to complete.

I am happy to provide additional context and specifics if needed.

**Comment by davidducos** [comment]
Hi @billynoah,
Can you try with --serialized-table-creation on myloader?
I need the log on myloader.

**Comment by billynoah** [comment]
Thanks - I'll give it a try.  Is there a way to send you logs privately?

**Comment by davidducos** [comment]
Hi @billynoah, 
Sure, you can send an email to me. 
get my email with
```
git log -1
```

**Comment by davidducos** [comment]
Hi @billynoah,
You can use -q 1 to avoid having large transactions that could cause those issues. Anyways will be really interesting to know what database version are you using to replicate it. 
Regards

**Comment by billynoah** [comment]
I will give that a whirl.  I did manager to get a log of a crash.  I email it to you.

Server version: 8.0.36-0ubuntu0.20.04.1 (Ubuntu)

**Comment by davidducos** [comment]
Hi @billynoah the log that you share shows more data threads than the ones that you configure in myloader under -t. 
Can you share the exact command that you run?

**Comment by billynoah** [comment]
Here is the exact command I ran which generated that log:

myloader --defaults-extra-file=$HOME/.my.cnf  --verbose=3
--logfile=myloader.log  -o -t 8 -B muscle_dev -d muscle_live

**Comment by billynoah** [comment]
Here is the exact command (dbname and dirname redacted)

```
myloader --defaults-extra-file=$HOME/.my.cnf  --verbose=3 --logfile=myloader.log  -o -t 8 -B dbname -d dirname
```

**Comment by davidducos** [comment]
Hi @billynoah 
In the command that you share, you are not using `--serialized-table-creation`

**Comment by davidducos** [comment]
Hi @billynoah, 
According to the latest log that you send me, there is an issue on the database. Is the database crashing? can you check running myloader with less threads?

**Comment by billynoah** [comment]
Yes the database crashes when this happens.  Not otherwise.  I have tried with less threads and it doesn't seem to eliminate the issue.  Can you point to specifically what you saw in the log that might help me debug this issue further?

**Comment by davidducos** [comment]
If the database is crashing, it might be running out of resources, how many cores has it? why don't to check the error log to see what is causing the crash?

**Comment by billynoah** [comment]
Yes, it's running out of resources. But my question is why? This only happens with myloader and I am getting crashes even with only 2 threads. The most I can tell you is that at some point there is a process "waiting for handler commit" that takes very long and consumes huge amounts of ram.  When this happens, the server runs out of RAM and OOM kills mysqld.

I just ran myloader and here is a screenshot of one of the processes where it was waiting for handler commit for 95 seconds. Based on the logs I sent , do you have any idea what causes this "waiting for handler commit"?

![Screenshot 2024-07-11 at 2 31 11 PM](https://github.com/mydumper/mydumper/assets/3309911/8ea45155-d0c8-4f43-8bfa-94584fd64e14)

To answer your question: I have the same issue on three different servers.  One of them, our production server has 64 GB of RAM and 16 cores.

**Comment by davidducos** [comment]
Did you try with -q, --queries-per-transaction = 1 as I told you before? as seems that the transactions are too long. It might be possible that you are hitting a bug.

**Comment by billynoah** [comment]
Yes I did try that.  I may not have tried both `q 1` and `--serialized-table-creation` together. I will try a few times.

I guess I am more interested in trying to understand the underlying cause than experimenting with command line options. I was hoping the log file would give us more clear answers.

**Comment by davidducos** [comment]
myloader is not checking the status of the server during the import, we might implement that in the future, that is why there is no valuable information on the logs about current issue. I think that you might be able to find something more useful in PMM if you are using it to monitor the server status.

**Comment by billynoah** [comment]
I just ran this command:
```
 myloader --defaults-extra-file=$HOME/.my.cnf  --verbose=3 --logfile=myloader.log  --serialized-table-creation -q 1 -o -t 8 -B dbname -d dirname
 ```

The DB crashed / out of memory:
```
Jul 11 15:22:57 unk kernel: [3811952.354839] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/system.slice/mysql.service,task=mysqld,pid=1720298,uid=112
Jul 11 15:22:57 unk kernel: [3811952.355074] Out of memory: Killed process 1720298 (mysqld) total-vm:20334280kB, anon-rss:15852772kB, file-rss:0kB, shmem-rss:0kB, UID:112 pgtables:32820kB oom_score_adj:0
Jul 11 15:22:58 unk systemd[1]: mysql.service: Main process exited, code=killed, status=9/KILL
Jul 11 15:22:58 unk systemd[1]: mysql.service: Failed with result 'signal'.
Jul 11 15:22:58 unk systemd[1]: mysql.service: Scheduled restart job, restart counter is at 5.
Jul 11 15:22:58 unk systemd[1]: Stopped MySQL Community Server.
Jul 11 15:22:58 unk systemd[1]: Starting MySQL Community Server...
Jul 11 15:23:10 unk systemd[1]: Started MySQL Community Server.
```

I put a pretty heavy load on this server regularly and it never crashes.  It's hard to believe this isn't somehow related to myloader.  I will continue to try more things - like single threaded, but at that point I don't think there will be any performance benefit.

**Comment by davidducos** [comment]
Are you running myloader and the database in the same server?
can you try to run myloader in a different server?

**Comment by billynoah** [comment]
Yes they are on the same server.  I highly doubt that will affect anything
but I'll give it a go when I have time and update this ticket. Thanks for
all the time you've spent looking at this.

**Comment by davidducos** [comment]
Hi @billynoah, 
any feedback about this issue? I'm closing it until we get more info.

**Comment by billynoah** [comment]
Not sure @davidducos - I'm on version:
```
mydumper v0.16.7-5, built against MySQL 8.0.37-29 with SSL support
```

In my automation I've been using:
```
--disable-redo-log --serialized-table-creation -q 500 -t 8
```
No issues recently. Thanks for your help.
